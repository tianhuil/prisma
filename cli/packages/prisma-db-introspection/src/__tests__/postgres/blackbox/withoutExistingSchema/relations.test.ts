import testSchema from '../common'

describe('Introspector', () => {
  test('relation with relation table', async () => {
    await testSchema(`CREATE TABLE product (
      id         serial PRIMARY KEY  -- implicit primary key constraint
    , product    text NOT NULL
    );
    
    CREATE TABLE bill (
      id       serial PRIMARY KEY
    , bill     text NOT NULL
    );
    
    CREATE TABLE bill_product (
      bill_id    int REFERENCES bill (id) ON UPDATE CASCADE ON DELETE CASCADE
    , product_id int REFERENCES product (id) ON UPDATE CASCADE
    );`)
  })

  test('relation with relation table with extra column', async () => {
    await testSchema(`CREATE TABLE product (
      id         serial PRIMARY KEY  -- implicit primary key constraint
    , product    text NOT NULL
    );
    
    CREATE TABLE bill (
      id       serial PRIMARY KEY
    , bill     text NOT NULL
    );
    
    CREATE TABLE bill_product (
      bill_id    int REFERENCES bill (id) ON UPDATE CASCADE ON DELETE CASCADE
    , product_id int REFERENCES product (id) ON UPDATE CASCADE
    , some_other_column text NOT NULL
    );`)
  })

  test('relation with inline relation column', async () => {
    await testSchema(`CREATE TABLE product (
      id           serial PRIMARY KEY  -- implicit primary key constraint
    , description  text NOT NULL
    );
    
    CREATE TABLE bill (
      id         serial PRIMARY KEY
    , bill       text NOT NULL
    , product_id int REFERENCES product (id) ON UPDATE CASCADE
    );`)
  })

  test('relation with inline relation column NOT NULL', async () => {
    await testSchema(`CREATE TABLE product (
      id           serial PRIMARY KEY  -- implicit primary key constraint
    , description  text NOT NULL
    );
    
    CREATE TABLE bill (
      id         serial PRIMARY KEY
    , bill       text NOT NULL
    , product_id int NOT NULL REFERENCES product (id) ON UPDATE CASCADE
    );`)
  })
})
