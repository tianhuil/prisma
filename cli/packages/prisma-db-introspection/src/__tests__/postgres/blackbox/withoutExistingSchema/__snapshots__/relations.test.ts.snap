// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Introspector relation with inline relation column 1`] = `
"type Bill @pgTable(name: \\"bill\\") {
  id: Int! @unique
  bill: String!
  product: Product @pgColumn(name: \\"product_id\\")
}

type Product @pgTable(name: \\"product\\") {
  id: Int! @unique
  bill: [Bill]
  description: String!
}"
`;

exports[`Introspector relation with inline relation column 2`] = `
"type Bill @db(name: \\"bill\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"bill_id_seq\\", initialValue: 1, allocationSize: 1)
  bill: String!
  product: Product @db(name: \\"product_id\\")
}

type Product @db(name: \\"product\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"product_id_seq\\", initialValue: 1, allocationSize: 1)
  bill: [Bill]
  description: String!
}"
`;

exports[`Introspector relation with inline relation column NOT NULL 1`] = `
"type Bill @pgTable(name: \\"bill\\") {
  id: Int! @unique
  bill: String!
  product: Product! @pgColumn(name: \\"product_id\\")
}

type Product @pgTable(name: \\"product\\") {
  id: Int! @unique
  bill: [Bill]
  description: String!
}"
`;

exports[`Introspector relation with inline relation column NOT NULL 2`] = `
"type Bill @db(name: \\"bill\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"bill_id_seq\\", initialValue: 1, allocationSize: 1)
  bill: String!
  product: Product! @db(name: \\"product_id\\")
}

type Product @db(name: \\"product\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"product_id_seq\\", initialValue: 1, allocationSize: 1)
  bill: [Bill]
  description: String!
}"
`;

exports[`Introspector relation with relation table 1`] = `
"type Bill @pgTable(name: \\"bill\\") {
  id: Int! @unique
  bill: String!
  billProduct: [BillProduct]
}

type BillProduct @pgTable(name: \\"bill_product\\") @relationTable {
  bill: Bill @pgColumn(name: \\"bill_id\\")
  product: Product @pgColumn(name: \\"product_id\\")
}

type Product @pgTable(name: \\"product\\") {
  id: Int! @unique
  billProduct: [BillProduct]
  product: String!
}"
`;

exports[`Introspector relation with relation table 2`] = `
"type Bill @db(name: \\"bill\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"bill_id_seq\\", initialValue: 1, allocationSize: 1)
  bill: String!
  billProduct: [BillProduct]
}

type BillProduct @db(name: \\"bill_product\\") @relationTable {
  bill: Bill @db(name: \\"bill_id\\")
  product: Product @db(name: \\"product_id\\")
}

type Product @db(name: \\"product\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"product_id_seq\\", initialValue: 1, allocationSize: 1)
  billProduct: [BillProduct]
  product: String!
}"
`;

exports[`Introspector relation with relation table with extra column 1`] = `
"type Bill @pgTable(name: \\"bill\\") {
  id: Int! @unique
  bill: String!
  billProduct: [BillProduct]
}

type BillProduct @pgTable(name: \\"bill_product\\") {
  bill: Bill @pgColumn(name: \\"bill_id\\")
  product: Product @pgColumn(name: \\"product_id\\")
  someOtherColumn: String! @pgColumn(name: \\"some_other_column\\")
}

type Product @pgTable(name: \\"product\\") {
  id: Int! @unique
  billProduct: [BillProduct]
  product: String!
}"
`;

exports[`Introspector relation with relation table with extra column 2`] = `
"type Bill @db(name: \\"bill\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"bill_id_seq\\", initialValue: 1, allocationSize: 1)
  bill: String!
  billProduct: [BillProduct]
}

type BillProduct @db(name: \\"bill_product\\") {
  bill: Bill @db(name: \\"bill_id\\")
  product: Product @db(name: \\"product_id\\")
  someOtherColumn: String! @db(name: \\"some_other_column\\")
}

type Product @db(name: \\"product\\") {
  id: Int! @id(strategy: SEQUENCE) @sequence(name: \\"product_id_seq\\", initialValue: 1, allocationSize: 1)
  billProduct: [BillProduct]
  product: String!
}"
`;
